### PR TITLE
Define planning process

### PR DESCRIPTION
**Pull request by nicksnyder**

### CODE REVIEW

File: doc/dev/proposals/2018-12-18-introduce-proposal-process.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+Other reading:
+- [Scaling Engineering Teams via Writing Things Down and Sharing - aka RFCs](https://blog.pragmaticengineer.com/scaling-engineering-teams-via-writing-things-down-rfcs/)
+- [Go proposal template](https://github.com/golang/proposal/blob/master/design/TEMPLATE.md)
+- [Rust RFC template](https://github.com/rust-lang/rfcs/blob/master/0000-template.md)
+
+## Proposal
+
+Adopt the proposal process documented in [README.md](README.md).
+
+## Rational
```
Comment (beyang) [inline doc/dev/proposals/2018-12-18-introduce-proposal-process.md]:
```suggestion
## Rationale
```

File: doc/dev/proposals/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Proposal process
+
+Changes to our architecture are proposed, discussed, and approved according to the following process:
```
Comment (beyang) [inline doc/dev/proposals/README.md]: Just architecture changes or any non-trivial feature development?

Comment (nicksnyder) [reply doc/dev/proposals/README.md]: The latter (although it is hard to imagine a non-trivial feature that wouldn't also require a change in architecture). I will update the language to clarify.

File: doc/dev/proposals/2018-12-18-introduce-proposal-process.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+This is a proposal to define a proposal process.
+
+## Background
+
+Our engineering planning process has evolved over time.
+
+From 2017-08 to 2018-09, we checked in proposals to [sourcegraph/docs-private/](https://github.com/sourcegraph/docs-private/).
+- We were on a monthly planning and release cycle (1 week planning, 2 weeks build, 1 week endgame and testing).
+- It was hard for plan N projects in one week and get them reviewed and approved by the smaller number of appropriate owners/reviewers.
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply doc/dev/proposals/2018-12-18-introduce-proposal-process.md]: You can use the suggestion feature to directly suggest a change now:

```suggestion
- It was hard to plan N projects in one week and get them reviewed and approved by the smaller number of appropriate owners/reviewers.
```

File: doc/dev/proposals/2018-12-18-introduce-proposal-process.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+Other reading:
+- [Scaling Engineering Teams via Writing Things Down and Sharing - aka RFCs](https://blog.pragmaticengineer.com/scaling-engineering-teams-via-writing-things-down-rfcs/)
+- [Go proposal template](https://github.com/golang/proposal/blob/master/design/TEMPLATE.md)
+- [Rust RFC template](https://github.com/rust-lang/rfcs/blob/master/0000-template.md)
+
+## Proposal
+
+Adopt the proposal process documented in [README.md](README.md).
+
+## Rationale
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply doc/dev/proposals/2018-12-18-introduce-proposal-process.md]: Here is how I think about it (and I am open to suggestion on how to make this more clear in the template).

**Background** What is the problem that is being solved? What motivates this proposal?
**Proposal** A proposed solution.
**Rationale** A discussion of alternate solutions and tradeoffs, and why the proposed solution was ultimately chosen.

File: doc/dev/proposals/0000-00-00-template.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+## Rational
+
+[A discussion of alternate approaches and the trade offs, advantages, and disadvantages of the specified approach.]
+
+## Implementation
+
+[A description of the steps in the implementation, who will do them, and when, in the form of a checklist.]
+
+### Test plan
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply doc/dev/proposals/0000-00-00-template.md]: added

File: doc/dev/proposals/0000-00-00-template.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+[A very brief explanation of this proposal.]
+
+## Background
+
+[An introduction of the necessary background and the problem being solved by the proposed change.]
+
+## Proposal
+
+[A precise statement of the proposed change.]
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply doc/dev/proposals/0000-00-00-template.md]: I think this falls under the release plan section that I added.

File: doc/dev/proposals/2018-12-18-introduce-proposal-process.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    - Planning for a project could take as long (or as short) as necessary.
+- We decided to capture plans in [GitHub issues](https://github.com/sourcegraph/sourcegraph/issues?q=is%3Aopen+is%3Aissue+label%3Aroadmap) and require more explicit checklists (which could be dynamically updated to communicate progress).
+
+In December 2018 we onboarded two teammates and realized three pain points:
+1. GitHub issues are not the ideal medium to iterate on a design.
+    - If you continuously update the original issue description to reflect the best known design, then it makes it hard to follow the subsequent conversation on the issue.
+    - It is possible to see the previous states of the issue description in the GitHub UI, but it isn't obvious and it isn't coupled with the conversation that motivated those changes.
+    - [Example](https://github.com/sourcegraph/sourcegraph/issues/1467)
+2. GitHub issues are not an ideal store of important design decisions because it is mixed with other things like bugs.
+3. We have out of date documentation for services. It is generally hard to keep separate documentation up to date, so as a first step we should focus on documenting changes.
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply doc/dev/proposals/2018-12-18-introduce-proposal-process.md]: This is certainly true but feels beside the point. Our previous planning processes and this proposal all rely on writing things down because we think that is important, so we aren't really changing anything here. All we are really debating/changing is "what is the best medium for writing things down and discussing".

File: doc/dev/proposals/2018-12-18-introduce-proposal-process.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+Adopt the proposal process documented in [README.md](README.md).
+
+## Rationale
+
+This is the next step in the evolution of our planning process.
+- We want to re-capture the benefits of the old process (written, searchable, record of markdown files for why we did things) now that we no longer have the same constraints that caused us to switch away from the old docs-private process (i.e. get all planning for the month done in 1 week).
+- We want to maintain the benefits of the current process:
+  - Explicit checklist for action items serves as a proof of understanding what needs to happen.
+  - Checklist is dyncamically updated to communicate the current state of the project.
```
Comment (nicksnyder) [inline doc/dev/proposals/2018-12-18-introduce-proposal-process.md]:
```suggestion
  - Checklist is dynamically updated to communicate the current state of the project.
```

**Comment by beyang** [comment]
I am writing a proposal doc for the Java extension now, and one thing I notice missing from the previous format was a place for "known risks / uncertainties". You could shove these into the "Implementation" section, but maybe it's worth having an explicit section/subsection in the template (I'm going to include one in the Java proposal).

**Comment by nicksnyder** [comment]
> "known risks / uncertainties"

Yeah, this is worth noting, but I think it falls under "Rational" (i.e. risks and tradeoffs should be discussed in the context of the next best alternative).

> instruct the author to include milestone dates

Agree, will update.

File: doc/dev/proposals/README.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# Proposal process
+
+Changes to our product or architecture are proposed, discussed, and approved according to the following process:
+
+1. Propose a change by opening a PR to add a new document to this directory that follows the [template](0000-00-00-template.md). The file name has the following components:
+  - The date the proposal was first authored (YYYY-MM-DD).
+  - The dash (`-`) separated title.
+  - The markdown (`.md`) suffix.
+2. Request a review from your manager and other relevant owners. The reviews are expected to respond to new and updated proposals within 48 hours (excluding weekends and holidays). It is appropriate to ping reviewers after 24 hours. The entire process should not take longer than one week. If a proposal is not approved within a week, the manager will make a final judgement call whether a project is approved as written or rejected.
```
Comment (lguychard) [inline doc/dev/proposals/README.md]:
```suggestion
 2. Request a review from your manager and other relevant owners. The reviewers are expected to respond to new and updated proposals within 48 hours (excluding weekends and holidays). It is appropriate to ping reviewers after 24 hours. The entire process should not take longer than one week. If a proposal is not approved within a week, the manager will make a final judgement call whether a project is approved as written or rejected.
```

File: doc/dev/proposals/0000-00-00-template.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+## Done date
+
+[The date that all goals of the project will be accomplished. This is a firm commitment that we can communicate to anyone (e.g. internal team, external customers). As such, the date should account for all factors that could impact the it (including, but not limited to, necessary refactoring, writing automated tests, fixing bugs discovered, etc.), discounting by probability. Once approved, this date cannot be moved without agreement from the project approvers. If a project is ever at risk of not hitting the planned date, the project owner should immediately bring this to the attention of the approvers and also do whatever work is necessary to bring the project back on schedule. Descoping is generally preferred to moving the date. If the date needs to be moved, it is better to move it early in the project's lifecycle, so prioritize work that gives you more information about the remaining work first.]
+
+## Retrospective
+
+[This section is completed after the project is completely done (i.e. the checklist is complete).]
```
Comment (lguychard) [inline doc/dev/proposals/0000-00-00-template.md]: I feel like this template may be missing a KPI/success indicators section -- completing a checklist of engineering tasks does not necessarily make a project a success. Rationale is the "why" of a project, and Checklist the "how", but before committing to starting a project we should also think about "what will make this a success", and assess whether the completed product meets those goals in Retrospective.

This is especially relevant for larger projects, but may be overkill for smaller, very technical proposals.

Comment (nicksnyder) [reply doc/dev/proposals/0000-00-00-template.md]: Agree, added

File: doc/dev/plans/company-goals.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+# Company goals
+
+## 2018 Q1
+
+Increase the number of companies that get from 1 to 20 daily users.
```
Comment (nicksnyder) [inline doc/dev/plans/company-goals.md]: @sqs @beyang for now, this is where I want to track these goals. We can move this later if there is a better location.

**Comment by nicksnyder** [comment]
After iterating on this a bit I am happy with the current state. We will discuss this process in January after everyone is back.

Please add any comments/questions/concerns as comments on this diff.

File: doc/dev/plans/company-goals.md (hunk 11 above)
Comment (nicksnyder) [reply doc/dev/plans/company-goals.md]: @dadlerj too
